### PR TITLE
Fix Aggregation For Visualizing Large Data

### LINKED ISSUE #80: Add scale legend
Add a scale legend to show user what the on-cell color encoding signifies

### LINKED ISSUE #128: Disconnected Supernodes
Currently all the datasets we have are connected for supernode aggregation and all nodes that do not have any neighbors (nodes, supernodes) are removed if they have no neighbors. This causes a potential issue when we have a disconnected supernode such as in this particular case.

_Originally posted by @waxlamp in https://github.com/multinet-app/multimatrix/pull/120#discussion_r521503927_

### LINKED ISSUE #201: Fix Matrix Visualization when viewing child rows
Connections between child nodes are not visualized when aggregate row is expanded. They should be visualized like this from graffinity

image

### LINKED ISSUE #210: Change forEach to map for Aggregation.ts
There are places in aggregation.ts that can use map instead of foreach

### PR DESCRIPTION
**Pull request by mozartfish**
This PR fixes bugs that appear in the les mis data that are not visible in the small flight dataset.

Fixed Node Ordering for Aggregating by Group for Les Mis
Close #210 
Close #201 
Close #128 
Close #80 

Opened the following issues for this PR
Open #213 
Open #212 
Open #211 

image

Demonstration of Changes
https://user-images.githubusercontent.com/24800816/109598833-f360ef80-7ace-11eb-92f0-66697f649ecc.mp4

### CODE REVIEW

**Comment by gotdairyya** [comment]
Can you have this branch target to your dev branch rather than master?

**Review by JackWilb — COMMENTED** [review]
Would you please make a note of which issues you opened because of these changes?

File: src/lib/aggregation.ts [hunk 1]
```diff
@@ -223,18 +291,19 @@ function expandSuperNodeData(
     });
 
     // Add a parent position value for the child nodes
-    nodeCopy.forEach((node: Node) => {
+    const expandedNodes = nodeCopy.map((node: Node) => {
       if (node.type === 'childnode') {
         const parentNodeID = superChildrenMap.get(node.id);
-        if (parentNodeID !== undefined) {
+        if (parentNodeID) {
```
Comment (JackWilb) [inline src/lib/aggregation.ts:297]: Same as above

File: src/lib/aggregation.ts [hunk 2]
```diff
@@ -345,21 +384,27 @@ export function expandSuperNetwork(
     superChildrenDict,
   );
 
-  // Create a new set of neighbors for the new network nodes
   let neighborNodes: Node[] = [];
-  if (expandNodes !== undefined && expandLinks !== undefined) {
+  if (expandNodes && expandLinks) {
```
Comment (JackWilb) [inline src/lib/aggregation.ts:388]: Same as above

File: src/lib/aggregation.ts [hunk 3]
```diff
@@ -372,12 +417,12 @@ function retractSuperNodeData(
...
   // If the supernode exists in the dictionary,
   // get the children of the supernode
-  if (superNode !== undefined) {
+  if (superNode) {
     const superChildrenIDs = superNode.CHILDREN;
     const childNodes: Node[] = [];
     superChildrenIDs.forEach((id: string) => {
       const childNode = childrenNodeNameDict.get(id);
-      if (childNode !== undefined) {
+      if (childNode) {
```
Comment (JackWilb) [inline src/lib/aggregation.ts:425]: A couple more to fix

File: src/lib/aggregation.ts [hunk 4]
```diff
@@ -472,7 +514,7 @@ export function retractSuperNetwork(
 
   // Create a new set of neighbors for the new network nodes
   let neighborNodes: Node[] = [];
-  if (retractNodes !== undefined && retractLinks !== undefined) {
+  if (retractNodes && retractLinks) {
```
Comment (JackWilb) [inline src/lib/aggregation.ts:517]: The last 2 😄

**Review by JackWilb — CHANGES_REQUESTED** [review]
Generally it looks good. Just a few instance of the same issue that should be really easy to clean up

File: src/lib/aggregation.ts (hunk 4 above)
Comment (mozartfish) [reply src/lib/aggregation.ts:517]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts (hunk 3 above)
Comment (mozartfish) [reply src/lib/aggregation.ts:425]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts (hunk 2 above)
Comment (mozartfish) [reply src/lib/aggregation.ts]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts (hunk 1 above)
Comment (mozartfish) [reply src/lib/aggregation.ts]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts [hunk 5]
```diff
@@ -203,12 +271,12 @@ function expandSuperNodeData(
...
   // If the supernode exists in the dictionary,
   // get the children of the supernode
-  if (superNode !== undefined) {
+  if (superNode) {
     const superChildrenIDs = superNode.CHILDREN;
     const childNodes: Node[] = [];
     superChildrenIDs.forEach((id: string) => {
       const childNode = childrenNodeNameDict.get(id);
-      if (childNode !== undefined) {
+      if (childNode) {
```
[the remark this answers is not in the document]
Comment (mozartfish) [reply src/lib/aggregation.ts]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts [hunk 6]
```diff
@@ -203,12 +271,12 @@ function expandSuperNodeData(
 
   // If the supernode exists in the dictionary,
   // get the children of the supernode
-  if (superNode !== undefined) {
+  if (superNode) {
```
[the remark this answers is not in the document]
Comment (mozartfish) [reply src/lib/aggregation.ts]: resolved with this commit https://github.com/multinet-app/multimatrix/pull/208/commits/6dfbea5a134cba6d71d69051872c0e1d559f7788

File: src/lib/aggregation.ts (hunk 3 above)
Comment (JackWilb) [reply src/lib/aggregation.ts:425]: This doesn't look to be fixed

File: src/lib/aggregation.ts (hunk 4 above)
Comment (JackWilb) [reply src/lib/aggregation.ts:517]: This doesn't look to be fixed

**Review by JackWilb — CHANGES_REQUESTED** [review]

**Review by JackWilb — APPROVED** [review]

**Pull request merged**